// locomotive-site-builder-rendering/lib/shared/components/html.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HTMLComponent = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _base = require('base-64');

var _base2 = _interopRequireDefault(_base);

var _page = require('./page');

var _page2 = _interopRequireDefault(_page);

var _fonts = require('./fonts');

var _asset_path = require('../asset_path');

var _asset_path2 = _interopRequireDefault(_asset_path);

var _url_to = require('../url_to');

var _url_to2 = _interopRequireDefault(_url_to);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function encodePreloadedData(data) {
  return _base2.default.encode(encodeURIComponent(JSON.stringify(data)));
}

function HTMLComponent(props) {
  var _props$data = props.data,
      site = _props$data.site,
      page = _props$data.page,
      options = _props$data.options;

  var options = options || {};
  var Layout = props.definitions.layout;
  var assetPath = (0, _asset_path2.default)(props.assetHost, site.theme_checksum);
  var urlTo = (0, _url_to2.default)(options);
  var clientOptions = { env: props.env, assetHost: props.assetHost };

  return _react2.default.createElement(
    Layout,
    {
      site: site,
      page: page,
      assetPath: assetPath,
      options: options,
      contentEditing: options.content_editing },
    _react2.default.createElement(
      'div',
      { id: 'locomotive-site-builder-page-anchor' },
      _react2.default.createElement(_page2.default, {
        site: site,
        page: page,
        options: options,
        definitions: props.definitions.blocks,
        assetPath: assetPath,
        urlTo: urlTo,
        contentEditing: options.content_editing })
    ),
    _react2.default.createElement('script', { src: assetPath('/themes/' + site.theme_handle + '/site_builder.js'), type: 'text/javascript' }),
    _react2.default.createElement('script', { type: 'text/javascript', dangerouslySetInnerHTML: { __html: '\n        window.SiteBuilder = ' + JSON.stringify(clientOptions) + ';\n        window.__PRELOADED__DATA__ = "' + encodePreloadedData(props.data) + '";\n        require("themes/' + site.theme_handle + '/frontend/javascripts/main");\n        require("themes/' + site.theme_handle + '/client").default.init();\n      ' } }),
    (0, _fonts.renderFonts)({ loadAllFonts: options.load_all_fonts })
  );
}

exports.HTMLComponent = HTMLComponent;
