// locomotive-site-builder-rendering/lib/shared/context.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
function buildContext(request, themes, attributes) {
  return new Promise(function (resolve, reject) {
    var context = Object.assign({
      path: null,
      status: null,
      rendered: false,
      params: request.params,
      themes: themes,
      externalData: {}
    }, attributes);

    if (context.path === null || context.path === undefined) context.path = 'index';

    context.mountedOn = context.previewEnabled && context.siteId ? '/_preview/' + context.siteId : '';

    context.isRendered = function () {
      return this.rendered === true;
    }.bind(context);

    context.setTemplate = function (name, data) {
      this.path = name;
      return this.setExternalData(data);
    }.bind(context);

    context.setExternalData = function (data) {
      this.externalData = data;
      return this;
    }.bind(context);

    context.redirectTo = function (location, type) {
      var _location = location + '';

      if (!/^https?:\/\//.test(location)) _location = this.mountedOn + location;

      // we don't want end users see the SiteBuiler official page
      if (location === 'https://sitebuilder.locomotivecms.com') _location = '/maintenance.html';

      this.location = _location;
      this.status = type || 301;
      this.rendered = true;

      return this;
    }.bind(context);

    context.render = function (body, status, contentType) {
      this.status = status || 200;
      this.body = body;
      this.contentType = contentType || 'text/html; charset=UTF-8';
      this.rendered = true;
      return this;
    }.bind(context);

    resolve(context);
  });
}

exports.default = buildContext;
